src/lightning/fabric/utilities/distributed.py

```diff
@@ -23,9 +27,67 @@ class group:  # type: ignore
         WORLD = None
 
 
+if TYPE_CHECKING:
+    from lightning.fabric.strategies import Strategy
+
+
 log = logging.getLogger(__name__)
 
 
+def is_shared_filesystem(strategy: "Strategy", path: Optional[_PATH] = None, timeout: int = 3) -> bool:
+    """Checks whether the filesystem under the given path is shared across all processes.
+
+    This function should only be used in a context where distributed is initialized.
+
+    Args:
+        strategy: The strategy being used, either from Fabric (``fabric.strategy``) or from Trainer
+            (``trainer.strategy``).
+        path: The path to check. Defaults to the current working directory. The user must have permissions to write
+            to this path or the parent folder, and the filesystem must be writable.
+        timeout: If any of the processes can't list the file created by rank 0 within this many seconds, the
+            filesystem is determined to be not shared.
+
+    """
+    # Fast path: Any non-local filesystem is considered shared (e.g., S3)
+    if path is not None and fsspec.utils.get_protocol(str(path)) != "file":
+        return True
+
+    path = Path(Path.cwd() if path is None else path).resolve()
+
+    # Fast path: Only distributed strategies can detect shared filesystems
+    if not hasattr(strategy, "world_size") or strategy.world_size == 1:
+        return True
+
+    # Fast path: If the path is not the same on all ranks or does not exist, we know it's not a shared filesystem
+    rank_zero_path = strategy.broadcast(path)
+    if not strategy.reduce_boolean_decision(rank_zero_path == path and path.exists(), all=True):
+        return False
+
+    path = path.parent if path.is_file() else path
+    check_file = path / ".lightning_shared_fs_check"
+    check_file.unlink(missing_ok=True)
+
+    strategy.barrier()
+    if strategy.is_global_zero:
+        # Rank 0 creates the file
+        check_file.touch()
+        found = True
+    else:
+        # All other ranks will wait until they find the file or timeout
+        start = time.perf_counter()
+        found = False
+        while not found and (time.perf_counter() - start) < timeout:
+            found = check_file.exists()
+    strategy.barrier()
+
+    all_found = strategy.reduce_boolean_decision(found, all=True)
+
+    with contextlib.suppress(OSError, PermissionError, FileNotFoundError):  # handle race condition on deletion
```

Comment:
When would the race condition trigger a PermissionError?

Both PermisionnError and FileNotFoundError are OSError subclasses anyways

```suggestion
    with contextlib.suppress(OSError):  # handle race condition on deletion
```